source/postprocess/visualization/maximum_horizontal_compressive_stress.cc

```diff
@@ -47,134 +47,161 @@ namespace aspect
+        else
+          comp_map = {{0u,0u},{1u,1u},{2u,2u},{0u,1u},{0u,2u},{1u,2u}};
+
+        auto pack_tensor = [&](const SymmetricTensor<2,dim> &t,
+                               Vector<double> &v)
+        {
+          Assert (v.size() == comp_map.size(), ExcInternalError());
+          for (unsigned k=0; k<v.size(); ++k)
+            {
+              const auto &p = comp_map[k];
+              v[k] = t[p.first][p.second];
+            }
+        };
+
+        auto unpack_tensor = [&](const Vector<double> &v,
+                                 SymmetricTensor<2,dim> &t)
+        {
+          Assert (v.size() == comp_map.size(), ExcInternalError());
+          t = SymmetricTensor<2,dim>(); // zero
+          for (unsigned k=0; k<v.size(); ++k)
+            {
+              const auto &p = comp_map[k];
+              t[p.first][p.second] = v[k];
+              if (p.first != p.second)
+                t[p.second][p.first] = v[k];
+            }
+        };
+
+        // --- STEP 1: compute compressive stress components at quadrature points
+        const unsigned int n_tensor_components = (dim==2 ? 3u : 6u);
+        std::vector<Vector<double>> stress_components(n_quadrature_points,
+                                                       Vector<double>(n_tensor_components));
+
         for (unsigned int q=0; q<n_quadrature_points; ++q)
           {
             const SymmetricTensor<2,dim> strain_rate = in.strain_rate[q];
             const SymmetricTensor<2,dim> deviatoric_strain_rate
               = (this->get_material_model().is_compressible()
-                 ?
-                 strain_rate - 1./3 * trace(strain_rate) * unit_symmetric_tensor<dim>()
-                 :
-                 strain_rate);
+                 ? strain_rate - 1./3 * trace(strain_rate) * unit_symmetric_tensor<dim>()
+                 : strain_rate);
 
             const double eta = out.viscosities[q];
 
-            // first compute the stress tensor, ignoring the pressure
-            // for the moment (the pressure has no effect on the
-            // direction since it just adds a multiple of the identity
-            // matrix to the stress, but because it is large, it may
-            // lead to numerical instabilities)
-            //
-            // note that the *compressive* stress is simply the
-            // negative stress
-            const SymmetricTensor<2,dim> compressive_stress = -2*eta*deviatoric_strain_rate;
-
-            // then find a set of (dim-1) horizontal, unit-length, mutually orthogonal vectors
+            SymmetricTensor<2,dim> compressive_stress = in.pressure[q] * unit_symmetric_tensor<dim>();
+
+            if (this->get_parameters().enable_elasticity)
+              {
+                const auto elastic_additional_out =
+                  out.template get_additional_output_object<MaterialModel::ElasticAdditionalOutputs<dim>>();
+
+                if (elastic_additional_out != nullptr)
```

Comment:
Again, not wrong, but I would suggest to use the same approach as in `postprocess/visualization/stress.cc` and assert that elastic additional outputs exists if `this->get_parameters().enable_elasticity` is true. I.e. if elasticity is enabled, the material model needs to provide elastic outputs, otherwise we do not allow to compute the stress, because it is likely wrong.

```
                Assert(elastic_additional_out != nullptr, ExcMessage("Elastic Additional Outputs are needed for the 'maximum horizontal compressive stress' postprocessor, but they have not been created."));
```

And then drop the inner if-condition.